File: 43-multiply-strings/43-multiply-strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if num1 == "0" or num2 == "0":
            return "0"
        
        res = [0] * 400
        for i1, d1 in enumerate(map(int, reversed(num1))):
            carry = 0
            for i2, d2 in enumerate(map(int, reversed(num2))):
                carry, d = divmod(d1 * d2 + carry, 10)
                c, res[i1 + i2] = divmod(res[i1 + i2] + d, 10) 
                carry += c
            if carry:
                res[i1 + i2 + 1] = carry
        
        leading_zeros = True
        ans = ""
        for d in reversed(res):
            if leading_zeros and d:
                leading_zeros = False
            if not leading_zeros:
                ans += str(d)
        return ans

File: 43-multiply-strings/43-multiply-strings.java
class Solution {
    public String multiply(String num1, String num2) {
        if(num1.equals("0") || num2.equals("0")){
            return "0";
        }
        
        if(num1.length() < num2.length()){
            String tmp = num1;
            num1 = num2;
            num2 = tmp;
        }
        
        int k = num1.length() + num2.length();
        int[] ans = new int[k];
        for(int j = num2.length() - 1; j >= 0; j--){
            k--;
            int d = num2.charAt(j) - '0';
            int carry = 0;
            int a = 0;
            for(int i = num1.length() - 1; i >= 0; i--){
                int p = (num1.charAt(i) - '0') * d + carry + ans[k - a];
                ans[k - a] = p % 10;
                carry = p / 10;
                a++;
            }
            while(carry > 0){
                int sum = ans[k - a] + carry;
                ans[k - a] = sum % 10;
                carry = sum / 10;
                a++;
            }
        }
        
        StringBuilder sb = new StringBuilder();
        boolean leading = true;
        for(int i = 0; i < ans.length; i++){
            if(!leading || ans[i] > 0){
                leading = false;
                sb.append((char) (ans[i] + 48));
            }
        }
        return sb.toString();
    }
}